### PR TITLE
Rewrite SIMD article as "SIMD and Hardware Intrinsics"

### PR DESCRIPTION
**Pull request by tannergooding**
Modernizes `docs/standard/simd.md`, which was out of date and missing many of the SIMD/vectorization concepts that modern .NET exposes. Retitled to **SIMD and Hardware Intrinsics** (the official term) and updated its toc entry.

The article now works through the layered SIMD support from highest to lowest level, aligned with the current dotnet/runtime vectorization guidance:

- `System.Numerics` fixed-shape vector/matrix types
- `Vector<T>` (variable-width)
- Cross-platform, fixed-width `Vector64/128/256/512<T>`
- Platform-specific hardware intrinsics (`X86`, `Arm`, `Wasm`)
- `TensorPrimitives` for ready-made vectorized span math

----------

Highlights:

- A near-complete **common-operations reference** table for the fixed-width vector APIs, grouped by category.
- **Code-structure and remainder-handling** guidance -- width fan-out, small-input jump tables, and masked vs. idempotent tail handling (with the overlap/`ConditionalSelect` reasoning).
- An **instruction-set configuration-knob reference** for .NET 11 (`EnableHWIntrinsic`, `MaxVectorTBitWidth`, `PreferredVectorBitWidth`, and the per-grouping `Enable*` x86/x64 and Arm64 switches), with an explicit callout that these are diagnostic tools, that they don''t affect ReadyToRun/Native AOT or necessarily internal runtime routines, that enabling an unsupported knob is ignored, and that the set may differ across releases.
- **Test all code paths** and **benchmark to confirm the win** sections, plus refreshed best practices leading with "reuse the higher-level APIs first" and "start with `Vector128<T>`".

----------

All code snippets live in a new `docs/standard/snippets/simd/csharp/` project (net11.0) and were built and run to verify output before inclusion. `bin/`/`obj/` are gitignored and not part of the change.

> [!NOTE]
> This article and its snippets were drafted with AI assistance (`ai-usage: ai-assisted`) and reviewed/verified against the runtime source and a compiled snippet project.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/fundamentals/toc.yml](https://github.com/dotnet/docs/blob/5fc5938a54538c19911e2ad42ca658a070e4c9be/docs/fundamentals/toc.yml) | [docs/fundamentals/toc](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/toc?branch=pr-en-us-54834) |
| [docs/standard/simd.md](https://github.com/dotnet/docs/blob/5fc5938a54538c19911e2ad42ca658a070e4c9be/docs/standard/simd.md) | [Use SIMD and hardware intrinsics in .NET](https://review.learn.microsoft.com/en-us/dotnet/standard/simd?branch=pr-en-us-54834) |

### CODE REVIEW

File: docs/standard/snippets/simd/csharp/SimpleVectorization.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+namespace SimdSnippets;
+
+public static class SimpleVectorization
+{
+    // <VectorizedRemainder>
+    // Idempotent search that re-processes the final vector instead of a scalar loop.
+    public static bool Contains(ReadOnlySpan<int> buffer, int searched)
+    {
+        Debug.Assert(Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<int>.Count);
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply docs/standard/snippets/simd/csharp/SimpleVectorization.cs]: Fixed -- added a scalar fallback for a buffer shorter than one vector, so it's correct for any length. The runner now covers the small and empty cases.

> [!NOTE]
> Reply posted by Copilot on my behalf.

File: docs/standard/simd.md [hunk 2]
```diff
@@ -1,121 +1,282 @@
...
+On Arm64, each logical grouping of instruction sets has its own switch:
+
+| Knob (`DOTNET_` prefix) | Default | Gates |
+| --- | --- | --- |
+| `EnableArm64Aes` | `1` | AES |
+| `EnableArm64Atomics` | `1` | Large System Extensions (LSE) atomics |
+| `EnableArm64Crc32` | `1` | CRC32 |
+| `EnableArm64Dczva` | `1` | `DC ZVA` cache-zeroing |
+| `EnableArm64Dp` | `1` | Dot Product |
+| `EnableArm64Rdm` | `1` | Rounding Double Multiply (RDMA) |
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply docs/standard/simd.md]: Fixed -- it's `RDM` (Rounding Doubling Multiply Accumulate); corrected to match the knob name.

> [!NOTE]
> Reply posted by Copilot on my behalf.

File: docs/standard/simd.md [hunk 3]
```diff
@@ -1,121 +1,282 @@
...
+| `EnableArm64Sha1` | `1` | SHA1 |
+| `EnableArm64Sha256` | `1` | SHA256 |
+| `EnableArm64Rcpc` | `1` | Release Consistent processor consistent (RCPC) |
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply docs/standard/simd.md]: Fixed -- dropped the duplicated word and corrected the casing to `RCpc`/`RCpc2`.

> [!NOTE]
> Reply posted by Copilot on my behalf.

File: docs/standard/simd.md [hunk 4]
```diff
@@ -1,121 +1,282 @@
...
+| Reorder | Rearrange elements by index or interleave two vectors | `Shuffle`, `Reverse`, `Zip`, `ZipLower`, `ZipUpper`, `Unzip`, `UnzipEven`, `UnzipOdd`, `ConcatLowerLower`, `ConcatLowerUpper`, `ConcatUpperLower`, `ConcatUpperUpper` |
+| Lane access | Read or replace individual elements and halves, or resize the vector | `GetElement`, `WithElement`, `ToScalar`, `GetLower`, `GetUpper`, `WithLower`, `WithUpper`, `ToVector256` |
+
+> [!TIP]
+> Several operations have `Estimate` and `Native` variants—for example, `MultiplyAddEstimate`, `ClampNative`, `MinNative`, `MaxNative`, `ShuffleNative`, and `ConvertToInt32Native`. They map to a faster hardware instruction that trades some precision or gives up an IEEE edge-case guarantee (such as NaN handling), so reach for them only when a benchmark shows the exact form is the bottleneck and the looser semantics are acceptable.
+
+> [!NOTE]
+> `Vector256.Shuffle` treats its input as a single 256-bit vector, where-as the platform-specific `Avx2.Shuffle` operates as two independent 128-bit lanes. The cross-platform API is the more portable choice, but confirm the behavior you need when porting hand-written intrinsics.
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply docs/standard/simd.md:121]: I can remove this if required, but it is rather a personal writing quirk of myself (the author) and such quirks are prevalent in books and text throughout history.

Trying to normalize everything is not strictly a positive, IMO.

File: docs/standard/simd.md [hunk 5]
```diff
@@ -1,121 +1,282 @@
...
+| `EnableAPX` | `0` | APX (extended general-purpose registers) |
+| `EnableAES` | `1` | AES, PCLMULQDQ |
+| `EnableAVX512VP2INTERSECT` | `1` | AVX-512 VP2INTERSECT |
+| `EnableAVXIFMA` | `1` | AVX-IFMA |
+| `EnableAVXVNNI` | `1` | AVX-VNNI |
+| `EnableAVXVNNIINT` | `1` | VEX AVX-VNNI-INT8 and AVX-VNNI-INT16 |
+| `EnableGFNI` | `1` | GFNI |
+| `EnableSHA` | `1` | SHA |
+| `EnableVAES` | `1` | VAES, VPCLMULQDQ |
+| `EnableWAITPKG` | `1` | WAITPKG |
```
Comment (teo-tsirpanis) [inline docs/standard/simd.md:232]: Some instruction sets like these, are not yet available as intrinsics.

Comment (tannergooding) [reply docs/standard/simd.md:232]: This is somewhat intentional. I'm 100% going to forget to update this to be in sync all the time, so I'd rather give out the full list including the ones we already have detection support for and will be lighting up in .NET 12.

File: docs/fundamentals/toc.yml [hunk 6]
```diff
@@ -411,7 +411,7 @@ items:
                 href: ../standard/memory-and-spans/index.md
               - name: Memory<T> and Span<T> usage guidelines
                 href: ../standard/memory-and-spans/memory-t-usage-guidelines.md
-          - name: SIMD-enabled types
+          - name: SIMD and Hardware Intrinsics
```
Comment (gewarren) [inline docs/fundamentals/toc.yml]:
```suggestion
          - name: SIMD and hardware intrinsics
```

File: docs/standard/simd.md [hunk 7]
```diff
@@ -1,121 +1,282 @@
...
+description: Learn about the layered SIMD support in .NET, from System.Numerics vector types to cross-platform Vector64/128/256/512 APIs, hardware intrinsics, and TensorPrimitives.
 author: FIVIL
 ms.author: tagoo
-ms.date: 03/26/2026
+ms.date: 07/18/2026
 ai-usage: ai-assisted
 ---
 
-# Use SIMD-accelerated numeric types
+# Use SIMD and Hardware Intrinsics in .NET
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
# Use SIMD and hardware intrinsics in .NET
```

File: docs/standard/simd.md [hunk 8]
```diff
@@ -1,121 +1,282 @@
...
+| API | Namespace | When to use it |
+| --- | --- | --- |
+| Fixed-purpose vector and matrix types | <xref:System.Numerics?displayProperty=fullName> | Graphics and geometry math with 2-4 element vectors, matrices, quaternions, and planes. |
+| `Vector<T>` | <xref:System.Numerics?displayProperty=fullName> | Portable, variable-width vectorization when you don't need per-platform control. |
+| `Vector64<T>`, `Vector128<T>`, `Vector256<T>`, `Vector512<T>` | <xref:System.Runtime.Intrinsics?displayProperty=fullName> | Cross-platform, fixed-width vectorization with fine-grained control. This is the recommended starting point for new vectorized algorithms. |
+| Hardware intrinsics | <xref:System.Runtime.Intrinsics.X86?displayProperty=fullName>, <xref:System.Runtime.Intrinsics.Arm?displayProperty=fullName>, <xref:System.Runtime.Intrinsics.Wasm?displayProperty=fullName> | Specific processor instructions that the higher-level APIs don't expose, for the last bit of performance on a hot path. |
+| `TensorPrimitives` | <xref:System.Numerics.Tensors?displayProperty=fullName> | Ready-made, vectorized math over spans. It does the vectorization for you. |
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
| Fixed-purpose vector and matrix types | <xref:System.Numerics> | Graphics and geometry math with 2-4 element vectors, matrices, quaternions, and planes. |
| `Vector<T>` | <xref:System.Numerics> | Portable, variable-width vectorization when you don't need per-platform control. |
| `Vector64<T>`, `Vector128<T>`, `Vector256<T>`, `Vector512<T>` | <xref:System.Runtime.Intrinsics> | Cross-platform, fixed-width vectorization with fine-grained control. This is the recommended starting point for new vectorized algorithms. |
| Hardware intrinsics | <xref:System.Runtime.Intrinsics.X86>, <xref:System.Runtime.Intrinsics.Arm?displayProperty=fullName>, <xref:System.Runtime.Intrinsics.Wasm?displayProperty=fullName> | Specific processor instructions that the higher-level APIs don't expose, for the last bit of performance on a hot path. |
| `TensorPrimitives` | <xref:System.Numerics.Tensors> | Ready-made, vectorized math over spans. It does the vectorization for you. |
```

Namespaces are always displayed fully qualified.

File: docs/standard/simd.md [hunk 9]
```diff
@@ -1,121 +1,282 @@
...
+
+- <xref:System.Numerics.Vector.IsHardwareAccelerated?displayProperty=nameWithType> reports whether `Vector<T>` operations are accelerated.
+- <xref:System.Runtime.Intrinsics.Vector128.IsHardwareAccelerated?displayProperty=nameWithType>, and the `Vector64`/`Vector256`/`Vector512` equivalents, report acceleration for each fixed width.
+
+These properties are turned into constants by the JIT, so the branches you don't take are eliminated and there's no runtime cost to checking them. Don't cache the values; read them directly where you need them. The same applies to the `Count` properties (for example, `Vector128<T>.Count`), which are also JIT-time constants.
+
+Most operations on an accelerated width are themselves accelerated, but it isn't guaranteed for every operation. For example, floating-point division may be accelerated where integer division isn't. When `Vector256` is accelerated, `Vector128` usually is too, but there's no guarantee, so check each width you use.
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
Most operations on an accelerated width are themselves accelerated, but it isn't guaranteed for every operation. For example, floating-point division might be accelerated where integer division isn't. When `Vector256` is accelerated, `Vector128` usually is too, but there's no guarantee, so check each width you use.
```

File: docs/standard/simd.md [hunk 10]
```diff
@@ -1,121 +1,282 @@
...
+
+> [!TIP]
+> If an operation you need isn't accelerated on a platform you care about, or you'd like a new cross-platform API, file an issue on [dotnet/runtime](https://github.com/dotnet/runtime/issues). The same applies to codegen improvements.
+
+Not every element type is valid for every vector. `Vector128<T>` and its siblings support the primitive numeric types (`byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `nint`, and `nuint`) today, and that set may grow to include other types in the future. Use `Vector128<T>.IsSupported` to determine whether a given `T` is valid, which is especially useful from generic code.
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
Not every element type is valid for every vector. `Vector128<T>` and its siblings support the primitive numeric types (`byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `nint`, and `nuint`) today, and that set might grow to include other types in the future. Use `Vector128<T>.IsSupported` to determine whether a given `T` is valid, which is especially useful from generic code.
```

File: docs/standard/simd.md [hunk 11]
```diff
@@ -1,121 +1,282 @@
...
+
+Each width has a generic type (`Vector128<T>`) for the data and a non-generic static class (<xref:System.Runtime.Intrinsics.Vector128>) that holds most of the operations, including static factory methods like `Create` and `Load`. Operators such as `+`, `&`, and `<<` are the idiomatic way to express arithmetic and bit operations; prefer them over the named method equivalents to avoid operator-precedence bugs and improve readability. For algorithms that depend on byte order, branch on <xref:System.BitConverter.IsLittleEndian>, which the JIT also folds to a constant.
+
+> [!NOTE]
+> On x86/x64, `Vector256<T>` operations are generally treated as two independent 128-bit "lanes". For most element-wise operations this is transparent, but operations that cross lanes (such as shuffles or pairwise/horizontal operations) may behave differently or cost more than the `Vector128` equivalent. Confirm with benchmarks before assuming a wider vector is faster.
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
> On x86/x64, `Vector256<T>` operations are generally treated as two independent 128-bit "lanes". For most element-wise operations this is transparent, but operations that cross lanes (such as shuffles or pairwise/horizontal operations) might behave differently or cost more than the `Vector128` equivalent. Confirm with benchmarks before assuming a wider vector is faster.
```

File: docs/standard/simd.md [hunk 12]
```diff
@@ -1,121 +1,282 @@
...
+
+### Common operations
+
+`Vector128` and its wider siblings expose a large API surface. You don't need to memorize it—know the categories and look up the details when you need them. Every operation has a software fallback for platforms that can't accelerate it. The table below covers essentially the whole surface.
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
`Vector128` and its wider siblings expose a large API surface. You don't need to memorize it—know the categories and look up the details when you need them. Every operation has a software fallback for platforms that can't accelerate it. The following table covers essentially the whole surface.
```

File: docs/standard/simd.md [hunk 13]
```diff
@@ -1,121 +1,282 @@
...
+
+:::code language="csharp" source="./snippets/simd/csharp/AdvancedVectorization.cs" id="VectorSmall":::
+
+`sizeof(T)` is also a JIT-time constant (and, with the `unmanaged` constraint, doesn't require an `unsafe` context in C# 15 / .NET 11), so `SumVectorSmall` dispatches on the element width to a table sized for the widest vector's worth of elements—the same approach <xref:System.Numerics.Tensors.TensorPrimitives> uses. Only the 4-byte table is shown; the 1-, 2-, and 8-byte tables share its shape. Its larger cases fold the leftover with a `Vector256` or `Vector128` using two overlapping loads—one from the start, one from the end—so the wider-remainder handling for the omitted `Vector512`/`Vector256` paths lives right in the jump table. The two loads overlap whenever the length isn't an exact multiple of the width, so the tail is masked to the additive identity with `ConditionalSelect` before it's summed. That mask is only needed because addition is non-idempotent; an idempotent operation such as a search could fold the overlapping tail in directly. A buffer on hardware with no vectorization at all falls through to `SumScalar`, an ordinary scalar loop.
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
`sizeof(T)` is also a JIT-time constant (and, with the `unmanaged` constraint, doesn't require an `unsafe` context in C# 15 / .NET 11), so `SumVectorSmall` dispatches on the element width to a table sized for the widest vector's worth of elements—the same approach <xref:System.Numerics.Tensors.TensorPrimitives> uses. Only the 4-byte table is shown; the 1-byte, 2-byte, and 8-byte tables share its shape. Its larger cases fold the leftover with a `Vector256` or `Vector128` using two overlapping loads—one from the start, one from the end—so the wider-remainder handling for the omitted `Vector512`/`Vector256` paths lives right in the jump table. The two loads overlap whenever the length isn't an exact multiple of the width, so the tail is masked to the additive identity with `ConditionalSelect` before it's summed. That mask is only needed because addition is non-idempotent; an idempotent operation such as a search could fold the overlapping tail in directly. A buffer on hardware with no vectorization at all falls through to `SumScalar`, an ordinary scalar loop.
```

Comment (gewarren) [reply docs/standard/simd.md]: 1-byte, 2-byte, and 8-byte

File: docs/standard/simd.md [hunk 14]
```diff
@@ -1,121 +1,282 @@
...
+
+### Load and store vectors safely
+
+For most code, `Vector128.Create(span)` and <xref:System.Runtime.Intrinsics.Vector128.CopyTo*> are the simplest way to move data between a span and a vector, and the JIT keeps them efficient. When you need the lower-level load and store—for example, to walk a buffer by managed reference—prefer the <xref:System.Runtime.Intrinsics.Vector128.LoadUnsafe*> and <xref:System.Runtime.Intrinsics.Vector128.StoreUnsafe*> overloads that take a managed reference and an `nuint` element offset. Unlike the pointer-based `Load`/`Store` overloads, they don't require pinning the buffer, and unlike raw reference arithmetic they don't need you to manually advance a `ref`. Both alternatives are easy to get wrong in ways that introduce garbage-collector holes or access violations.
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
For most code, `Vector128.Create(span)` and <xref:System.Runtime.Intrinsics.Vector128.CopyTo*> are the simplest way to move data between a span and a vector, and the JIT keeps them efficient. When you need the lower-level load and store—for example, to walk a buffer by managed reference—prefer the <xref:System.Runtime.Intrinsics.Vector128.LoadUnsafe*> and <xref:System.Runtime.Intrinsics.Vector128.StoreUnsafe*> overloads that take a managed reference and an `nuint` element offset. Unlike the pointer-based `Load`/`Store` overloads, they don't require pinning the buffer, and unlike raw reference arithmetic, they don't need you to manually advance a `ref`. Both alternatives are easy to get wrong in ways that introduce garbage-collector holes or access violations.
```

File: docs/standard/simd.md [hunk 15]
```diff
@@ -1,121 +1,282 @@
...
+
+> [!IMPORTANT]
+> These are diagnostic tools, meant primarily for testing and validation—exercising each code path, reproducing a hardware-specific issue, or confirming a fallback. They aren't designed for general or production use, and they aren't a stability contract: the set below is what .NET 11 recognizes; earlier releases exposed a different set—the baseline and AVX-512 knobs in particular were reconfigured—so confirm the names against the runtime version you target.
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
> These are diagnostic tools, meant primarily for testing and validation—exercising each code path, reproducing a hardware-specific issue, or confirming a fallback. They aren't designed for general or production use, and they aren't a stability contract. The following set is what .NET 11 recognizes; earlier releases exposed a different set—the baseline and AVX-512 knobs in particular were reconfigured—so confirm the names against the runtime version you target.
```

File: docs/standard/simd.md [hunk 16]
```diff
@@ -1,121 +1,282 @@
...
+>
+> They also have limits on what they reach. Because they gate JIT decisions, they don't affect code that was already compiled ahead of time through ReadyToRun or Native AOT, and they don't necessarily affect internal routines the runtime and core libraries use themselves. Treat them as a way to steer your own JIT-compiled code, not a global off switch for an instruction set.
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
> These tools also have limits on what they reach. Because they gate JIT decisions, they don't affect code that was already compiled ahead of time through ReadyToRun or Native AOT, and they don't necessarily affect internal routines the runtime and core libraries use themselves. Treat them as a way to steer your own JIT-compiled code, not a global off switch for an instruction set.
```

File: docs/standard/simd.md [hunk 17]
```diff
@@ -1,121 +1,282 @@
...
-    int remaining = length % Vector<double>.Count;
+## Benchmark to confirm the win
 
-    for (int i = 0; i < length - remaining; i += Vector<double>.Count)
-    {
-        var v1 = new Vector<double>(left, i);
-        var v2 = new Vector<double>(right, i);
-        (v1 + v2).CopyTo(result, i);
-    }
+Vectorization adds complexity, so measure that it pays off before you keep it. Use [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet), and use the same environment variables above to compare the scalar, `Vector128`, and `Vector256` implementations in one run. BenchmarkDotNet's disassembly diagnoser can also emit the generated assembly, which is invaluable when tuning high-performance code.
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
Vectorization adds complexity, so measure that it pays off before you keep it. Use [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet), and use the same environment variables shown previously to compare the scalar, `Vector128`, and `Vector256` implementations in one run. BenchmarkDotNet's disassembly diagnoser can also emit the generated assembly, which is invaluable when tuning high-performance code.
```

File: docs/standard/simd.md [hunk 18]
```diff
@@ -1,121 +1,282 @@
...
+- Reach for the existing higher-level APIs first. `Span<T>`, `string`, LINQ, `TensorPrimitives`, and the tensor types already accelerate many common operations for you—don't hand-roll what's already optimized and tested.
+- Start with `Vector128<T>`; it's accelerated on the broadest set of hardware, and you don't need `Vector256<T>` to get a correct, portable implementation. Add wider widths and hardware intrinsics only for measured hot paths.
+- Check `IsHardwareAccelerated` and `Count` directly instead of caching them; the JIT turns them into constants.
+- Always handle the loop remainder, and account for overlapping source and destination buffers when storing.
+- Write edge-case tests first, then a scalar solution, then express that scalar logic with the vector APIs.
+- Test every code path (including access violations) and benchmark realistic input sizes before you commit to the added complexity.
+
+## See also
+
+- <xref:System.Runtime.Intrinsics?displayProperty=fullName>
```
Comment (gewarren) [inline docs/standard/simd.md]:
```suggestion
- <xref:System.Runtime.Intrinsics>
```

**Review by gewarren — APPROVED** [review]

**Comment by tannergooding** [comment]
Thanks @gewarren -- applied all of these in 03d8f40: sentence-cased the title/H1/toc, dropped the redundant `displayProperty=fullName` from the namespace xrefs (including one in the intro paragraph you didn't flag), and took the copyedits (may->might, `following table`, acronym spacing, and the callout punctuation).

> [!NOTE]
> Reply posted by Copilot on my behalf.

File: docs/standard/simd.md [hunk 19]
```diff
@@ -1,121 +1,282 @@
...
+
+:::code language="csharp" source="./snippets/simd/csharp/AdvancedVectorization.cs" id="VectorSmall":::
+
+`sizeof(T)` is also a JIT-time constant (and, with the `unmanaged` constraint, doesn't require an `unsafe` context in C# 15 / .NET 11), so `SumVectorSmall` dispatches on the element width to a table sized for the widest vector's worth of elements—the same approach <xref:System.Numerics.Tensors.TensorPrimitives> uses. Only the 4-byte table is shown; the 1-byte, 2-byte, and 8-byte tables share its shape. Its larger cases fold the leftover with a `Vector256` or `Vector128` using two overlapping loads—one from the start, one from the end—so the wider-remainder handling for the omitted `Vector512`/`Vector256` paths lives right in the jump table. The two loads overlap whenever the length isn't an exact multiple of the width, so the tail is masked to the additive identity with `ConditionalSelect` before it's summed. That mask is only needed because addition is non-idempotent; an idempotent operation such as a search could fold the overlapping tail in directly. A buffer on hardware with no vectorization at all falls through to `SumScalar`, an ordinary scalar loop.
```
Comment (BillWagner) [inline docs/standard/simd.md]: We (ECMA committee) introduced a spec bug when we added the `unmanaged` constraint that we recently fixed. See this PR: https://github.com/dotnet/csharpstandard/pull/1721.

The new relaxed rules for `sizeof(T)` are part of the new unsafe model. That should still need the `preview` language  version in C# 15 / .NET 11. Otherwise, only the named built in types are allowed for sizeof in a safe context. We should call that out in the first sentence. My suggestion splits it into 2 sentences because it gets harder to read:





```suggestion
`sizeof(T)` is also a JIT-time constant, so `SumVectorSmall` dispatches on the element width to a table sized for the widest vector's worth of elements—the same approach <xref:System.Numerics.Tensors.TensorPrimitives> uses. (When the new memory safety model is enabled, `sizeof(T)` expressions on a type parameter with the `unmanaged` constraint is allowed in safe code.) Only the 4-byte table is shown; the 1-byte, 2-byte, and 8-byte tables share its shape. Its larger cases fold the leftover with a `Vector256` or `Vector128` using two overlapping loads—one from the start, one from the end—so the wider-remainder handling for the omitted `Vector512`/`Vector256` paths lives right in the jump table. The two loads overlap whenever the length isn't an exact multiple of the width, so the tail is masked to the additive identity with `ConditionalSelect` before it's summed. That mask is only needed because addition is non-idempotent; an idempotent operation such as a search could fold the overlapping tail in directly. A buffer on hardware with no vectorization at all falls through to `SumScalar`, an ordinary scalar loop.
```

**Review by BillWagner — COMMENTED** [review]
There's one clarification that we need here.

File: docs/standard/simd.md (hunk 19 above)
Comment (tannergooding) [reply docs/standard/simd.md]: @BillWagner aren't the `sizeof(T)` relaxations part of the general change to make pointers no longer unsafe and are not part of the unsafe-v2 model that is in preview and requires opt-in?

Comment (BillWagner) [reply docs/standard/simd.md]: I checked with @333fred.  These should only light up with LangVersion.= preview, even after .NET 11 G.A.

**Review by BillWagner — APPROVED** [review]
